File: python/01_rbac_milvus.py
from pymilvus import MilvusClient

CLUSTER_ENDPOINT = "YOUR_CLUSTER_ENDPOINT"
TOKEN = "YOUR_CLUSTER_TOKEN"

# 1. Set up a Milvus client
client = MilvusClient(
    uri=CLUSTER_ENDPOINT,
    token=TOKEN 
)

# 2. Create a role
res = client.list_roles()
print(res)

# Output
#
# ["admin", "public"]



client.create_role(role_name="read_only")

res = client.list_roles()
print(res)

# Output
#
# ["admin", "public", "read_only"]



# 3. Grant permissions
client.grant_privilege(
    role_name="read_only",
    object_type="Global",
    object_name="*",
    privilege="DescribeCollection"
)

# 4. Describe the role
res = client.describe_role(role_name="read_only")

print(res)

# Output
#
# {
#     "role": "read_only",
#     "privileges": [
#         {
#             "object_type": "Global",
#             "object_name": "*",
#             "db_name": "default",
#             "role_name": "read_only",
#             "privilege": "DescribeCollection",
#             "grantor_name": "root"
#         }
#     ]
# }



# 5. Create a user
res = client.list_users()
print(res)

# Output
#
# ["root"]



client.create_user(
    user_name="alice",
    password="123456"
)

res = client.list_users()
print(res)

# Output
#
# ["alice", "root"]



# 6. Assign the role to the user
client.grant_role(
    user_name="alice",
    role_name="read_only"
)

# 7. Describe the user
res = client.describe_user(user_name="alice")

print(res)

# Output
#
# {
#     "user_name": "alice",
#     "roles": "(\"read_only\")"
# }



# 8. Revoke the role from the user
client.revoke_role(
    user_name="alice",
    role_name="read_only"
)

# 9. Drop the user
client.drop_user(user_name="alice")

# 10. Revoke permissions
client.revoke_privilege(
    role_name="read_only",
    object_type="Global",
    privilege="DescribeCollection",
    object_name="*"
)

# 11. Drop the role
client.drop_role(role_name="read_only")

File: node/01_rbac_milvus.js
const fs = require("fs")
const { MilvusClient, DataType, sleep } = require("@zilliz/milvus2-sdk-node")

const address = "YOUR_CLUSTER_ENDPOINT"
const token = "YOUR_CLUSTER_TOKEN"

async function main() {
    // 1. Connect to the cluster
    const client = new MilvusClient({address, token})

    // 2. Create a role
    var res = await client.listRoles()

    console.log(res.results)

    // Output
    // 
    // [
    //   { users: [], role: { name: 'admin' } },
    //   { users: [], role: { name: 'public' } }
    // ]
    // 

    res =await client.createRole({
        roleName: "read_only"
    })

    console.log(res.error_code)

    // Output
    // 
    // Success
    // 

    res = await client.listRoles()

    console.log(res.results)

    // Output
    // 
    // [
    //   { users: [], role: { name: 'admin' } },
    //   { users: [], role: { name: 'public' } },
    //   { users: [], role: { name: 'read_only' } }
    // ]
    // 

    // 3. Grant permissions
    res = await client.grantPrivilege({
        roleName: "read_only",
        object: "Global",
        objectName: "*",
        privilegeName: "DescribeCollection"
    })

    console.log(res.error_code)

    // Output
    // 
    // Success
    // 

    // 4. Describe the role
    res = await client.describeRole({
        roleName: "read_only"
    })

    console.log(res.results)

    // Output
    // 
    // [ { users: [], role: { name: 'read_only' } } ]
    // 

    // 5. Create a user
    res = await client.listUsers()

    console.log(res.usernames)

    // Output
    // 
    // [ 'root' ]
    // 

    res = await client.createUser({
        username: "alice",
        password: "123456"
    })

    console.log(res.error_code)

    // Output
    // 
    // Success
    // 

    res = await client.listUsers()

    console.log(res.user_names)

    // Output
    // 
    // undefined
    // 

    // 6. Assign a role to a user
    res = await client.grantRole({
        username: "alice",
        roleName: "read_only"
    })

    console.log(res.error_code)

    // Output
    // 
    // Success
    // 

    // 7. Describe the user
    res = await client.describeUser({
        username: "alice"
    })

    console.log(res.results)

    // Output
    // 
    // [ { roles: [ [Object] ], user: { name: 'alice' } } ]
    // 

    // 8. Revoke the role from the user
    res = await client.revokeRole({
        username: "alice",
        roleName: "read_only"
    })

    console.log(res.error_code)

    // Output
    // 
    // Success
    // 

    // 9. Drop the user
    res = await client.dropUser({
        username: "alice"
    })

    console.log(res.error_code)

    // Output
    // 
    // Success
    // 

    // 10. Revoke permissions
    res = await client.revokePrivilege({
        roleName: "read_only",
        object: "Global",
        objectName: "*",
        privilegeName: "DescribeCollection"    
    })

    console.log(res.error_code)

    // Output
    // 
    // Success
    // 

    // 11. Drop the role
    res = await client.dropRole({
        roleName: "read_only"
    })

    console.log(res.error_code)

    // Output
    // 
    // Success
    // 

    res = await client.listRoles()

    console.log(res.results)

    // Output
    // 
    // [
    //   { users: [], role: { name: 'admin' } },
    //   { users: [], role: { name: 'public' } }
    // ]
    // 

}

main()